pkg/kubelet/kubelet.go

```diff
@@ -2845,6 +2855,117 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+// 'kl.maxContainerCheckpoints'.
+func (kl *Kubelet) handleMaxContainerCheckpoints(
+	podUID types.UID,
+	podFullName,
+	containerName string,
+	options *runtimeapi.CheckpointContainerRequest,
+) error {
+
+	if kl.maxContainerCheckpoints <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// error out early if this is <= 0. This additional error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxContainerCheckpoints must be greater than 0")
+	}
+
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.maxContainerCheckpointsLock.Lock()
+	defer kl.maxContainerCheckpointsLock.Unlock()
+
+	checkpointDirectory, err := os.ReadDir(kl.getCheckpointsDir())
+	if err != nil {
+		return fmt.Errorf(
+			"error reading checkpoint archive directory %s: %w",
+			kl.getCheckpointsDir(),
+			err,
+		)
+	}
+
+	regexString := fmt.Sprintf(
+		`checkpoint-%s-%s-\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.*tar`,
+		podFullName,
+		containerName,
+	)
+
+	re, err := regexp.Compile(regexString)
+	if err != nil {
+		return fmt.Errorf(
+			"error compiling regex %s: %w",
+			regexString,
+			err,
+		)
+	}
+
+	var checkpointArchives = []string{}
+	for _, archive := range checkpointDirectory {
+		// only select checkpoint archives which have been created from the same
+		// namespace/pod/container combination
+		if re.MatchString(archive.Name()) {
+			checkpointArchives = append(
+				checkpointArchives,
+				filepath.Join(
+					kl.getCheckpointsDir(),
+					archive.Name()),
+			)
+		}
+	}
+
+	checkpointArchivesCounter := len(checkpointArchives)
+	if checkpointArchivesCounter <= int(kl.maxContainerCheckpoints) {
+		// Nothing to clean up, return early.
+		return nil
+	}
+
+	sort.Slice(checkpointArchives, func(i, j int) bool {
+		stats := make([]time.Time, 2)
+		for ind, archive := range []string{
+			checkpointArchives[i],
+			checkpointArchives[j],
+		} {
+			stat, _ := os.Stat(archive)
+			stats[ind] = stat.ModTime()
+		}
+		return stats[0].Before(stats[1])
+	})
+
+	for _, archive := range checkpointArchives {
+		klog.InfoS(
+			"More than maxContainerCheckpoints",
+			"container", containerName,
+			"numberOfCheckpoints", checkpointArchivesCounter,
+			"maxContainerCheckpoints", kl.maxContainerCheckpoints,
+		)
+		klog.InfoS(
+			"Deleting container checkpoint archive",
+			"container", containerName,
+			"archive", archive,
+		)
+		err = os.Remove(archive)
+		if err != nil {
+			// If the removal fails the checkpoint archive might have been
+			// deleted externally. This is not fatal but unexpected.
+			// Logging it as an error, but not returning an error.
+			klog.ErrorS(
+				err,
+				"Removal of checkpoint archive failed",
+				"container", containerName,
+				"archive", archive,
+			)
+		}
+		checkpointArchivesCounter = -1
+		if checkpointArchivesCounter <= int(kl.maxContainerCheckpoints) {
```

Comment:
If we sort `checkpointArchives` in a way that it starts with oldest files, we can simply remove every files from `checkpointArchives[:len(checkpontArchives)-kl.maxContainerCheckpoints]`

Reply:
also changed as suggested